### PR TITLE
refactor: migrate MyInfo functionality to TypeScript

### PR DESCRIPTION
**Pull request by mantariksh**
This PR extracts MyInfo functionality from combined SPCP/MyInfo modules, and places it in its own service written in TypeScript.

### Why a service and not a module?
I wrote this with the future PublicForm migration in mind. Currently, the two middlewares exported by the MyInfo controller (`addMyInfo` and `verifyMyInfoVals`) are middlewares in the form retrieval and submissions flow respectively. However, we want the logic in all these middlewares to eventually be combined into one function which calls several services. The intention here is for MyInfo to be one of those services, and for the logic in the current MyInfo controller functions to be moved into the PublicForm controllers in future.

### What about tests?
I have removed the old tests so that this PR builds successfully. The new tests will come in a separate PR and will be merged into this branch.

### CODE REVIEW

File: src/app/services/myinfo/myinfo.service.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+            meta: {
+              requestedAttributes: params.requestedAttributes,
+              eServiceId: params.singpassEserviceId,
+            },
+          },
+          error,
+        })
+        return isCircuitOpen
+          ? new CircuitBreakerError(logMessage)
+          : new FetchMyInfoError(logMessage)
```
Comment (liangyuanruo) [inline src/app/services/myinfo/myinfo.service.ts:138]: it'll be simpler to use an if-else statement here instead of having to perform ternary op on `isCircuitOpen` twice.

File: src/app/services/myinfo/myinfo.service.ts [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+  }
+
+  checkMyInfoHashes(
+    responses: ProcessedFieldResponse[],
+    hashes: IHashes,
+  ): ResultAsync<IHashes, HashingError | HashDidNotMatchError> {
+    // Filter twice to get the types to cooperate
+    const responsesWithHashes: VisibleMyInfoResponse[] = responses
+      .filter(hasMyInfoAnswer)
+      .filter((response) => !!hashes[response.myInfo.attr])
```
Comment (liangyuanruo) [inline src/app/services/myinfo/myinfo.service.ts:258]: i believe you won't have to filter twice if you convert `hasMyInfoAnswer` into a type guard

Comment (liangyuanruo) [reply src/app/services/myinfo/myinfo.service.ts:258]: wait, the function is already a type guard; this filter is to obtain the responses that have had their hashes saved.

Comment (mantariksh) [reply src/app/services/myinfo/myinfo.service.ts:258]: it already is a type guard! apparently using a filter function like `(element) => typeGuard(element) && otherFunction(element)` doesn't actually return the correct type...

File: src/app/services/myinfo/myinfo.util.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    case MyInfoAttribute.BillingAddress:
+    case MyInfoAttribute.MailingAddress:
+      return formatAddress(myInfoData[myInfoAttr])
+    default:
+      // Categorical data lookup
+      if (CATEGORICAL_DATA_DICT[myInfoAttr]) {
+        return get(
+          CATEGORICAL_DATA_DICT,
+          [myInfoAttr, myInfoData[myInfoAttr].value!, 'description'],
+          '',
```
Comment (liangyuanruo) [inline src/app/services/myinfo/myinfo.util.ts:66]: there hasn't been any change in business logic here, right?

**Review by liangyuanruo — APPROVED** [review]
lgtm, do test submitting for all MyInfo fields before merging

File: src/app/services/myinfo/myinfo.util.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+/**
+ * Formats MyInfo attribute as phone number
+ * @param phone
+ * @example +65 98654321
+ * @returns Phone number if phone.nbr exists. Else return empty string.
+ */
+const formatPhoneNumber = (phone: IMyInfoPhoneNo): string => {
+  if (!phone || !phone.nbr) return ''
+
+  return phone.prefix && phone.code && phone.nbr
```
Comment (tshuli) [inline src/app/services/myinfo/myinfo.util.ts]: is is necessary to check phone.nbr again since this is already checked in L29?

File: src/app/services/myinfo/myinfo.util.ts (hunk 3 above)
Comment (mantariksh) [reply src/app/services/myinfo/myinfo.util.ts:66]: I've rearranged the logic in a (hopefully) cleaner way, but it shouldn't change the behaviour. the old code was this:
```
    const attr = myInfoData[myInfoAttr]
    // Categorical data lookup	
    if (CATEGORICAL_DATA_DICT[myInfoAttr]) {	
      return _.get(	
        CATEGORICAL_DATA_DICT,	
        [myInfoAttr, attr.value, 'description'],	
        '',	
      )	
    }	

    // Not in category, return by case	
    switch (myInfoAttr) {	
      // Phone numbers	
      case 'mobileno':	
      case 'homeno':	
        return this.formatPhoneNumber(attr)	
      case 'regadd':	
      case 'billadd':	
      case 'mailadd':	
        return this.formatAddress(attr)	
      default:	
        return _.get(attr, 'value', '')	
    }
```

File: src/app/services/myinfo/myinfo.util.ts (hunk 4 above)
Comment (mantariksh) [reply src/app/services/myinfo/myinfo.util.ts]: Good question! Probably not - I just copied the existing code here, but we can remove that check

**Comment by mantariksh** [comment]
waiting to finish writing tests + test on staging before merging

File: src/app/services/myinfo/myinfo.errors.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import { ApplicationError } from '../../modules/core/core.errors'
+
+/**
+ * Circuit breaker is open
+ */
+export class CircuitBreakerError extends ApplicationError {
+  constructor(message?: string) {
```
Comment (karrui) [inline src/app/services/myinfo/myinfo.errors.ts:7]: nit: can simplify logic by instantiating message right in the constructor:

```ts
constructor(message = 'Circuit breaker tripped') {
  super(message)
}
```

File: src/app/services/myinfo/myinfo.service.ts [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+const logger = createLoggerWithLabel(module)
+const MyInfoHash = getMyInfoHashModel(mongoose)
+
+const MYINFO_STG_PREFIX = 'STG2-'
+const MYINFO_PROD_PREFIX = 'PROD2-'
+const MYINFO_DEV_BASE_URL = 'http://localhost:5156/myinfo/v2/'
+
+export class MyInfoService {
+  #myInfoClientBreaker: CircuitBreaker<[IPersonBasicRequest], IPersonBasic>
```
Comment (karrui) [inline src/app/services/myinfo/myinfo.service.ts:50]: nit: can add JSDocs to these child vars also?

File: src/app/services/myinfo/myinfo.service.ts [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+      myInfoClientMode === MyInfoClientMode.Staging
+        ? MYINFO_STG_PREFIX
+        : MYINFO_PROD_PREFIX
+    if (nodeEnv === Environment.Prod) {
+      myInfoGovClient = new MyInfoGovClient({
+        realm,
+        singpassEserviceId,
+        privateKey: fs.readFileSync(myInfoKeyPath),
+        appId: myInfoPrefix + singpassEserviceId,
+        mode: myInfoClientMode,
```
Comment (karrui) [inline src/app/services/myinfo/myinfo.service.ts:88]: Can simplify this since most of it is the same?

```ts
const myInfoGovClient = new MyInfoGovClient({
        realm,
        singpassEserviceId,
        privateKey: fs.readFileSync(myInfoKeyPath),
        appId: myInfoPrefix + singpassEserviceId,
        mode: nodeEnv === Environment.Prod ? myInfoClientMode : MyInfoClientMode.Dev,
        baseUrl: nodeEnv === Environment.Prod ? undefined : MYINFO_DEV_BASE_URL
})
```

not sure if that makes it clearer or not though haha

File: src/app/services/myinfo/myinfo.service.ts [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+   * @param myInfoData
+   * @param currFormFields
+   * @returns currFormFields with the MyInfo fields prefilled with data from myInfoData
+   */
+  prefillMyInfoFields(
+    myInfoData: IPersonBasic,
+    currFormFields: IFieldSchema[],
+  ): Result<IPossiblyPrefilledField[], never> {
+    return ok(
+      currFormFields.map((field) => {
```
Comment (karrui) [inline src/app/services/myinfo/myinfo.service.ts:164]: nit: a little hard to parse. Can assign it to a variable first, then only return at the end?

```ts
const prefilledFields = .....

return ok(prefilledFields)
```

File: src/app/services/myinfo/myinfo.service.ts [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+   * @throws error if an error occurred while hashing the values or updating the database
+   */
+  saveMyInfoHashes(
+    uinFin: string,
+    formId: string,
+    prefilledFormFields: IPossiblyPrefilledField[],
+  ): ResultAsync<IMyInfoHashSchema | null, HashingError | DatabaseError> {
+    const readOnlyHashPromises = hashFieldValues(prefilledFormFields)
+    return ResultAsync.fromPromise(
+      Bluebird.props<IHashes>(readOnlyHashPromises),
```
Comment (karrui) [inline src/app/services/myinfo/myinfo.service.ts:195]: This is the only place we use Bluebird. Is there any reason? Should we just use normal promises and then we can delete the Bluebird dependency

**Review by karrui — APPROVED** [review]
very nice and clean 😍

**Comment by mantariksh** [comment]
unfortunately MyInfo staging is still down so we can't test this on staging yet. will give them a call later today

File: src/app/services/myinfo/myinfo.service.ts (hunk 7 above)
Comment (mantariksh) [reply src/app/services/myinfo/myinfo.service.ts:88]: hm, arguable! personally I think the existing way is cleaner just because of the `baseUrl` line. the code you wrote won't work because `baseUrl` isn't a constructor argument, it's a class member that we're modifying from outside.

File: src/app/services/myinfo/myinfo.service.ts (hunk 9 above)
Comment (mantariksh) [reply src/app/services/myinfo/myinfo.service.ts:195]: yeah in fact I added another call to `Bluebird.props`, so now there are 2 instead of 1. I just think it makes a lot more sense than something like `Promise.all` because we care about mapping each attribute to a `Promise<boolean>` and knowing which ones resolve to true/false. If we use `Promise.all` then we have to do something clunky like `_.zip` to line up the attributes and the Promises, which is really unnecessary since we should be using a Map-like data structure in the first place.

File: src/app/services/myinfo/myinfo.service.ts (hunk 7 above)
Comment (karrui) [reply src/app/services/myinfo/myinfo.service.ts:95]: Got it

**Comment by mantariksh** [comment]
myinfo staging still not working
